src/dd_license_attribution/metadata_collector/strategies/npm_collection_strategy.py

```diff
@@ -0,0 +1,207 @@
+    SourceCodeManager,
+)
+from dd_license_attribution.metadata_collector.metadata import Metadata
+from dd_license_attribution.metadata_collector.project_scope import (
+    ProjectScope,
+)
+from dd_license_attribution.metadata_collector.strategies.abstract_collection_strategy import (
+    MetadataCollectionStrategy,
+)
+
+
+class NpmMetadataCollectionStrategy(MetadataCollectionStrategy):
+    def __init__(
+        self,
+        top_package: str,
+        source_code_manager: SourceCodeManager,
+        project_scope: ProjectScope,
+    ) -> None:
+        self.top_package = top_package
+        self.source_code_manager = source_code_manager
+        self.only_root_project = project_scope == ProjectScope.ONLY_ROOT_PROJECT
+        self.only_transitive = (
+            project_scope == ProjectScope.ONLY_TRANSITIVE_DEPENDENCIES
+        )
+
+    def _extract_all_dependencies(self, lock_data: Dict[str, Any]) -> Dict[str, str]:
+        all_deps: Dict[str, str] = {}
+
+        if "packages" not in lock_data:
+            logging.warning("No 'packages' key found in package-lock.json.")
+            return all_deps
+
+        packages = lock_data["packages"]
+        # Find the root package key
+        root_key = "" if "" in packages else "./" if "./" in packages else None
+        if root_key is None:
+            logging.warning("Root package not found in lockfile 'packages'.")
+            return all_deps
+
+        root_pkg = packages[root_key]
+        if "dependencies" in root_pkg:
+            all_deps.update(root_pkg["dependencies"])
+
+        self._extract_transitive_dependencies(packages, all_deps)
+        return all_deps
+
+    def _extract_transitive_dependencies(
+        self, packages: Dict[str, Any], all_deps: Dict[str, str]
+    ) -> None:
+
+        processed_packages = set()
+
+        new_deps_found = True
+        while new_deps_found:
+            new_deps_found = False
+            current_deps = list(all_deps.items())
+
+            for pkg_name, _ in current_deps:
+                if pkg_name in processed_packages:
+                    continue
+
+                node_modules_key = f"node_modules/{pkg_name}"
+                if node_modules_key in packages:
+                    pkg_data = packages[node_modules_key]
+                    if "dependencies" in pkg_data:
+                        for dep_name, dep_version in pkg_data["dependencies"].items():
+                            if dep_name not in all_deps:
+                                all_deps[dep_name] = dep_version
+                                new_deps_found = True
+
+                processed_packages.add(pkg_name)
+
+    def augment_metadata(self, metadata: List[Metadata]) -> List[Metadata]:
+        updated_metadata = metadata.copy()
+        source_code_ref = self.source_code_manager.get_code(self.top_package)
+        if not source_code_ref:
+            return updated_metadata
+        project_path = source_code_ref.local_full_path
```

Comment:
It may be a little restrictive to get only the root_project for lock. I can see how many projects may embed JS in a secondary directory for UX.

Reply:
Agreed. I will implement this in follow up PR. Do you have a good project that uses this for testing?